File: ProyectoCivitasJava/src/civitas/SorpresaPorJugador.java
/**
 * Proyecto Civitas
 * Sergio Hervás Cobo
 * Javier Suárez Lloréns
 * Programación y Diseño Orientado a Objetos
 * 2º B3 - Grado en Ingeniería Informática
 * E.T.S.I. Informática y de Telecomunicación
 * Universidad de Granada
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author sergio
 */
public class SorpresaPorJugador extends Sorpresa {

    private int valor;

    SorpresaPorJugador(int valor, String texto) {
        super(texto);
        this.valor = valor;
    }

    @Override
    void aplicarAJugador(int actual, ArrayList <Jugador> todos){
        if(jugadorCorrecto(actual, todos)){
            informe(actual, todos);
            Sorpresa pagar = new SorpresaPagarCobrar(valor*-1, "Pagar");
            Sorpresa cobrar = new SorpresaPagarCobrar(valor*(todos.size()-1), "Cobrar");
            for(int k = 0; k < todos.size(); k++){
                if(k != actual)
                    pagar.aplicarAJugador(k, todos); 
                else
                    cobrar.aplicarAJugador(k, todos);
            }
        
        }        
    }
    
    @Override
    public String toString(){
        String salida = "Sorpresa " +  valor + " Por Jugador";
        return salida;
    }
    
}

File: ProyectoCivitasRuby/lib/modelo/sorpresa_por_jugador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Civitas
  class SorpresaPorJugador < Sorpresa
    public_class_method :new 
    
    def initialize(valor, texto)
      super(texto)
      @valor = valor  
    end

    #@Override
    def aplicarajugador(actual, todos)
      if(jugador_correcto(actual, todos))
        informe(actual, todos)
        s1 = SorpresaPagarCobrar.new(-1*@valor, "PAGAR")
        s2 = SorpresaPagarCobrar.new(@valor, "COBRAR")  
        for k in (0..todos.size) do
          if (k != actual)
            s1.aplicarajugador(k, todos)
            k += 1
          end
        end
        s2.aplicarajugador(actual, todos)     
      end
    end
    
    def to_s
      salida = "{@texto}\nSorpresa #{@valor} Por Jugador"
    end
 
  end
end